baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousIdentityKey.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.AbstractAsKeyOfSetOrMap;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.util.Name;
+
+/**
+ * Warns that users should not have a {@link java.util.regex.Pattern} as a key to a Set or Map.
+ */
+@BugPattern(
+        name = "DangerousIdentityKey",
+        summary = "Key type does not override equals() and hashCode, so comparisons will be done on"
+                + " reference equality only. If neither deduplication nor lookup are needed,"
+                + " consider using a List instead. Otherwise, use IdentityHashMap/Set,"
+                + " or an Iterable/List of pairs.",
+        severity = SeverityLevel.WARNING)
+public final class DangerousIdentityKey extends AbstractAsKeyOfSetOrMap {
```

Comment:
Should we try to push this upstream to error-prone proper? I'm curious if there is a reason they haven't added something of this shape (though this is the type of thing drilled fairly well by `Effective Java`, I'd like to have 🤖 compiler assistance). 

I can add a test case to confirm that this catches the construction in places like stream collectors, as that's where I noticed one example.

Also worth noting that since we're building on `AbstractAsKeyOfSetOrMap` it is specific to the actual constructed map/set type (not the apparent type of say `Map<Pattern, String>`) and analyzes the typical JDK & Guava (minus `WeakHash(Map|Set)`) hash based maps & sets where `hashCode` and `equals` are relevant. Not sure if we need/want a similar check for Caffeine caches & sorted/navigable map & set types. Reminds me of some places I've seen [`java.net.URL` as keys](http://errorprone.info/bugpattern/URLEqualsHashCode) 😢 .

https://github.com/google/error-prone/blob/c601758e81723a8efc4671726b8363be7a306dce/core/src/main/java/com/google/errorprone/bugpatterns/AbstractAsKeyOfSetOrMap.java#L43-L91

Reply:
That's a good point, this should be a strong candidate for an upstream contribution, although I've never had much luck with error-prone contributions, I can reach out again.

We could extend AbstractAsKeyOfSetOrMap to include caches (guava + caffeine) as well as collectors. I can PR something here shortly, and make a similar contribution to the abstract class upstream.